File: python/examples/cc0mon-api-get-owner.py
"""cc0mon-api-get-owner — fetch the current holder address for a token.

Usage::

    python cc0mon-api-get-owner.py --id 1
"""
from __future__ import annotations

import argparse
import dataclasses
import json
import sys

from cc0mon_sdk import (
    Client,
    ClientApiError,
    NetworkError,
    RateLimitError,
    ServerApiError,
    ValidationError,
    configure_logger,
)

ACTION = "get-owner"


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Fetch the current holder address for a cc0mon token.")
    parser.add_argument("--id", dest="token_id", type=int, required=True, help="Token id (1-10000)")
    args = parser.parse_args(argv)

    logger = configure_logger(ACTION)
    logger.info("script start id=%d", args.token_id)

    try:
        with Client() as client:
            owner_info = client.get_owner(args.token_id)
    except ValidationError as e:
        logger.error("validation: %s", e)
        print(f"validation error: {e}", file=sys.stderr)
        return 5
    except NetworkError as e:
        logger.error("network: %s", e)
        print(f"network error: {e}", file=sys.stderr)
        return 2
    except RateLimitError as e:
        logger.error("rate limited http %d: %s", e.status_code, e.body[:200])
        print(f"rate limited HTTP {e.status_code}: {e.body[:200]}", file=sys.stderr)
        return 3
    except ClientApiError as e:
        logger.error("http %d: %s", e.status_code, e.body[:200])
        print(f"client error HTTP {e.status_code}: {e.body[:200]}", file=sys.stderr)
        return 3
    except ServerApiError as e:
        logger.error("http %d: %s", e.status_code, e.body[:200])
        print(f"server error HTTP {e.status_code}: {e.body[:200]}", file=sys.stderr)
        return 4

    print(json.dumps(dataclasses.asdict(owner_info), indent=2, default=str))
    logger.info("script exit code=0 owner=%s", owner_info.owner)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: java/examples/cc0mon-api-get-owner.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 17
//DEPS com.cc0mon:sdk:0.1.0
//DEPS com.fasterxml.jackson.core:jackson-databind:2.17.0

import com.cc0mon.sdk.Client;
import com.cc0mon.sdk.Errors;
import com.cc0mon.sdk.LoggingSetup;
import com.cc0mon.sdk.Models.OwnerInfo;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

import java.util.logging.Logger;

class GetOwner {
    public static void main(String[] args) {
        Logger log = LoggingSetup.configure("get-owner");
        Integer id = parseId(args);
        if (id == null) { System.err.println("usage: cc0mon-api-get-owner.java --id <1..10000>"); System.exit(5); }
        log.info("script start id=" + id);
        try (Client client = new Client()) {
            OwnerInfo info = client.getOwner(id);
            System.out.println(new ObjectMapper().enable(SerializationFeature.INDENT_OUTPUT).writeValueAsString(info));
            log.info("script exit code=0 owner=" + info.owner());
        } catch (Errors.ValidationException e) { log.severe("validation: " + e.getMessage()); System.err.println("validation error: " + e.getMessage()); System.exit(5);
        } catch (Errors.NetworkException e)    { log.severe("network: " + e.getMessage());    System.err.println("network error: " + e.getMessage()); System.exit(2);
        } catch (Errors.ClientApiException e)  { log.severe("http " + e.getStatusCode());     System.err.println("client error HTTP " + e.getStatusCode() + ": " + e.getBody()); System.exit(3);
        } catch (Errors.ServerApiException e)  { log.severe("http " + e.getStatusCode());     System.err.println("server error HTTP " + e.getStatusCode() + ": " + e.getBody()); System.exit(4);
        } catch (Exception e) { log.severe("unexpected: " + e); System.err.println("unexpected: " + e); System.exit(1); }
    }

    static Integer parseId(String[] args) {
        for (int i = 0; i + 1 < args.length; i++) {
            if ("--id".equals(args[i])) try { return Integer.parseInt(args[i + 1]); } catch (NumberFormatException e) { return null; }
        }
        return null;
    }
}
